Dependabot docs (#617)
cagov/data-infrastructure: docs/code/dependabot-runbook.md

# Dependabot at CalData

> GitHub's [Dependabot](https://github.com/dependabot) service provides automated security scanning

Reviewer: ```suggestion
[Dependabot](https://github.com/dependabot) provides automated security scanning of dependencies in repositories hosted in GitHub.
```

This makes the previously very hard to read sentence more readable. It also reduces redundancy of "GitHub".

of dependencies to repositories hosted in GitHub.
It is enabled by default in the `cagov` organization.

## Why Dependabot?

Automated security scans, combined with practices like cooldowns, allow for a relatively

> painless way to keep on top of security vulnerabilities in our software dependencies.

Reviewer: ```suggestion
Automated security scans can reduce vulnerabilities in our software dependencies.
```

I removed cooldowns because you didn't link to it and we usually want to link to a definition at first mention. Also improves readability, no longer a very hard to read sentence.

## Decisions

1. Group pull requests. This instructs Dependabot to create a single grouped pull request
    for all package security upgrades, rather than a separate pull request for each individual
    security upgrade. We want to enable this grouping to reduce the noise from the alerts
    as a mitigation for [alarm fatigue](https://en.wikipedia.org/wiki/Alarm_fatigue).
1. Use dependency cooldowns. [Dependency cooldowns](https://docs.astral.sh/uv/concepts/resolution/#dependency-cooldowns)

>     are a very simple, but seemingly very effective, way to avoid pulling in compromised

Reviewer: ```suggestion
are a simple and effective way to avoid compromised
```

    dependencies. Our primary package manager, `uv`, supports them as a top-level configuration.

> 1. Avoid specific Dependabot configuration: we try to avoid Dependabot configuration

Reviewer: ```suggestion
1. Avoid specific Dependabot configuration. We avoid Dependabot configuration
```

>     via `dependabot.yml` files for operational simplicity. If we ever decide we need
>     to include such configuration we will incorporate it into our infrastructure template
>     so that all of our repositories share a similar process.

Reviewer: ```suggestion
    via `dependabot.yml` files for operational simplicity. If we need to include such configuration we do so in our infrastructure template so that all repositories share a similar process.
```

Author: I'm skipping this one because I think the rewrite says something different

## Repository setup

These steps only need to be done once per repository.

**1. Enable grouped pull requests:**
Dependabot is already enabled by default in all `cagov` repositories.

> However, we still need to enable pull request grouping to avoid getting one

Reviewer: ```suggestion
We still need to enable pull request grouping to avoid getting one
```

pull request per security upgrade. To enable this, go to
"Settings" → "Advanced Security" and enable "Grouped security updates".
An example grouped pull request is [here](https://github.com/cagov/data-infrastructure/pull/615).

**2. Enable dependency cooldowns:**
In a project `pyproject.toml`, ensure that the following configuration is set:
```toml
[tool.uv]
exclude-newer = "1 week"
```

> ## Runbook
> 
> There are two valid paths for responding to Dependabot alerts:
> you can either merge the auto-generated pull request, or create
> your own pull request that updates dependencies.
> Right now, we recommend the latter approach, as the former requires
> some work on our default CI/CD setup.

Reviewer: ```suggestion
There are two valid paths for responding to Dependabot alerts:

1. you can either merge the auto-generated pull request
2. create your own pull request that updates dependencies

Right now, we recommend the second approach. The first approach requires
some work on our default CI/CD setup.
```

### Creating your own pull request

1. Check out a new branch with a date slug in the branch name to disambiguate it

>     from other version upgrade PRs.

Reviewer: ```suggestion
1. Check out a new branch with a date in the branch name to set it apart from other related PRs.
```

    ```bash
    git switch main
    git pull <remote-name> main
    git switch -c security-bump-<YYYY>-<MM>-<DD>
    ```
1. Regenerate the `uv` lockfile with
    ```bash
    uv lock --upgrade
    ```
    This should bring in security
    fixes. It will *also* bring in version bumps for other packages. In some cases these
    incidental patches might *also* be a vector for supply chain attacks. This risk is
    mitigated by using cooldowns as discussed above.
1. Commit and push:
    ```bash

>     git commit -m "bump versions"

Reviewer 2: should we include git add uv.lock before git commit ? the current commands do not stage the lockfile change.

Author: Good catch!

    git push <remote-name> security-bump-<YYYY>-<MM>-<DD>
    ```
1. Open a pull request against `main` with your branch in GitHub. Once CI passes,
    feel free to merge. We don't need to rely on code review for security patches.

> 

Reviewer 2: I think adding a step to confirm the Dependabot alert is resolved after the updated lockfile is merged might help?

Author: I agree

### Merging Dependabot PRs

> This workflow currently does not function smoothly because, by default,
> Dependabot *does not* have access to repository secrets, and therefore
> CI/CD that does things like connect to Snowflake will not work.

Reviewer: ```suggestion
This workflow currently does not function smoothly because, by default. Dependabot *does not* have access to repository secrets so CI/CD that does things like connect to Snowflake will not work.
```

A long term fix to this could be the following:

1. Set up GitHub deploy environments for dev and prod.
2. Set up OIDC for GitHub actions to connect to, e.g., Snowflake from the deploy environment.
2. Allow for maintainer approval for PRs coming from services like Dependabot

Once those are configured, maintainers could simply click "Allow"
on Dependabot security bump PRs, and then CI would run on the Dependabot

> PR directly, followed by merge if it passes.

Reviewer: ```suggestion
Once those are configured, maintainers can click "Allow" on Dependabot security bump PRs, then CI will run on the Dependabot PR directly, followed by merge if it passes.
```

# Dependabot at CalData

GitHub's [Dependabot](https://github.com/dependabot) provides automated security scanning
of dependencies in repositories hosted in the platform.
It is enabled by default in the `cagov` organization.

## Why Dependabot?

Automated security scans can reduce vulnerabilities in our software dependencies.

> ## Decisions

Reviewer 3: for clarity -- these are decisions we've made, or decisions still up for discussion?

Author: Let's talk about it! I didn't hear much objection to them when I proposed them, but I wouldn't consider these "decisions made" until we have some consensus and they are merged!

Reviewer 3: My take so far:

1. Group pull requests -- I don't know if we have to protect against alarm fatigue if we're also implementing / enforcing a teamwide process for addressing them. Could go either way on this
2. Dependency cooldowns -- I think we should do this!
3. Avoid specific Dependabot configuration -- Why is it simpler to avoid configurations (assuming we would then apply all config as a default to our template? Is it because we have so many package ecosystems to monitor and it's not worth defining config for each one (as far as we know right now)?

Author: > Avoid specific Dependabot configuration -- Why is it simpler to avoid configurations (assuming we would then apply all config as a default to our template? Is it because we have so many package ecosystems to monitor and it's not worth defining config for each one (as far as we know right now)?

Basically, because while propagating changes from our template to downstream projects *works*, it's kind of annoying (somewhat evidenced by how infrequently we do it). So my motivation here was to try the simplest thing that could possibly work (no configuration) first, and then re-evaluate if we need to.

Author: @summer-mothwood are you okay with the proposed process as it sits?

    via `dependabot.yml` files for operational simplicity. If we ever decide we need
    to include such configuration we will incorporate it into our infrastructure template
    so that all of our repositories share a similar process.

## Repository setup

These steps only need to be done once per repository.

**1. Enable grouped pull requests:**
Dependabot is already enabled by default in all `cagov` repositories.
We still need to enable pull request grouping to avoid getting one
pull request per security upgrade. To enable this, go to
"Settings" → "Advanced Security" and enable "Grouped security updates".
An example grouped pull request is [here](https://github.com/cagov/data-infrastructure/pull/615).

**2. Enable dependency cooldowns:**
In a project `pyproject.toml`, ensure that the following configuration is set:
```toml
[tool.uv]
exclude-newer = "1 week"
```

## Runbook

There are two valid paths for responding to Dependabot alerts:

1. You can merge the auto-generated pull request
2. You can create your own pull request that updates dependencies.

Right now, we recommend the latter approach, as the former requires
some work on our default CI/CD setup.

> ### Creating your own pull request

Reviewer 3: ```suggestion
### Option 1 Creating your own pull request
```

Reviewer 3: numbering these 2 sections will help tie them back up to the list above. Want to make it abundantly clear that this is an either/or.

1. Check out a new branch with a date slug in the branch name to distinguish it
    from other version upgrade PRs.
    ```bash
    git switch main
    git pull <remote-name> main
    git switch -c security-bump-<YYYY>-<MM>-<DD>
    ```
1. Regenerate the `uv` lockfile with
    ```bash
    uv lock --upgrade
    ```
    This should bring in security
    fixes. It will *also* bring in version bumps for other packages. In some cases these
    incidental patches might *also* be a vector for supply chain attacks. This risk is
    mitigated by using cooldowns as discussed above.
1. Commit and push:
    ```bash
    git add uv.lock
    git commit -m "bump versions"
    git push <remote-name> security-bump-<YYYY>-<MM>-<DD>
    ```
1. Open a pull request against `main` with your branch in GitHub. Once CI passes,
    feel free to merge. We don't need to rely on code review for security patches.
1. Once your pull request is merged, the existing Dependabot PRs should detect it
    and auto-close.

> ### Merging Dependabot PRs

Reviewer 3: ```suggestion
### Option 2 Merging Dependabot PRs
```
